cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,449 @@
+    // just return original byte[] from the passed input stream. Trying to compute frames require loading all the
+    // dependent classes for a class and this might lead to class loading problems if some dependent classes are not
+    // available. For example HbaseTableUtil use Hbase classes and in in-memory and standalone cdap hbase classes
+    // are not present.
+    ClassWriter cw = new ClassWriter(0);
+
+    // SKIP_CODE to make the first pass faster since in the first pass we just want to process annotations to check
+    // if the class has any method with AuthEnforce annotation. If such method is found we also store the parameters
+    // which has the named annotations as specified in the entities field of the AuthEnforce.
+    cr.accept(new AuthEnforceAnnotationVisitor(className, cw), ClassReader.SKIP_CODE);
+
+    if (!methodAnnotations.isEmpty()) {
+      // We found some method which has AuthEnforce annotation so we need a second pass in to rewrite the class
+      cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES); // in second pass we COMPUTE_FRAMES and visit classes with
+      // EXPAND_FRAMES
+      cr.accept(new AuthEnforceAnnotationRewriter(className, cw), ClassReader.EXPAND_FRAMES);
+      return cw.toByteArray();
+    }
+    // if no AuthEnforce annotation was found then return the original class bytes
+    return classBytes;
+  }
+
+
+  /**
+   * A {@link ClassVisitor} which is used in the first pass of the {@link AuthEnforceRewriter} to detect and store
+   * method with {@link AuthEnforce} annotations. Note: This is a visitor, to know the order of in which the below
+   * overridden methods will be called and their overall responsibility please see {@link ClassVisitor} documentation.
+   */
+  private final class AuthEnforceAnnotationVisitor extends ClassVisitor {
+
+    private final String className;
+    private boolean interfaceClass;
+
+    AuthEnforceAnnotationVisitor(String className, ClassWriter cw) {
+      super(Opcodes.ASM5, cw);
+      this.className = className;
+    }
+
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+      // visit the header of the class to identify if this class is an interface
+      super.visit(version, access, name, signature, superName, interfaces);
+      interfaceClass = Modifier.isInterface(access);
+    }
+
+    @Override
+    public MethodVisitor visitMethod(final int access, final String methodName, final String methodDesc,
+                                     final String signature, final String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, methodName, methodDesc, signature, exceptions);
+
+      // No need to visit the class to look for AuthEnforce annotation if this class is an interface or the method is
+      // static.
+      if (interfaceClass || Modifier.isStatic(access)) {
+        return mv;
+      }
+
+      final boolean[] hasAuthEnforce = new boolean[1];
```

Comment:
Why not just make this (and the next) fields in the `MethodVisitor` you are creating below?